File: src/eval.py
import argparse
import csv
import datetime
import glob
import os
import sys
import torch

from model import PINN
from physics import HeatEquation1D, ViscousBurgers1D

EVALS_CSV = os.path.join("logs", "evals.csv")
EVALS_FIELDNAMES = ["name", "timestamp", "run_id", "max_abs_error", "l2_error"]


def evaluate(model, pde, device="cpu", grid=100):
    model.eval()
    with torch.no_grad():
        xs = torch.linspace(pde.x_min, pde.x_max, grid, device=device)
        ts = torch.linspace(0.0, pde.t_max, grid, device=device)
        X, Tm = torch.meshgrid(xs, ts, indexing="ij")
        x_flat = X.reshape(-1, 1)
        t_flat = Tm.reshape(-1, 1)

        u_pred = model(x_flat, t_flat)

        result = {
            "u_pred": u_pred,
            "X": X,
            "Tm": Tm,
        }

        u_exact = pde.analytical_solution(x_flat, t_flat)
        if u_exact is not None:
            max_err = (u_pred - u_exact).abs().max().item()
            l2_err = ((u_pred - u_exact) ** 2).mean().sqrt().item()
            result["u_exact"] = u_exact
            result["max_abs_error"] = max_err
            result["l2_error"] = l2_err

    return result


def resolve_run_path(arg):
    if arg is not None:
        return arg
    candidates = sorted(glob.glob(os.path.join("runs", "run_*.pt")))
    if not candidates:
        sys.exit("no runs found in runs/ — pass a path or run train.py first")
    return candidates[-1]


def load_run(run_path):
    if not os.path.exists(run_path):
        sys.exit(f"run file not found: {run_path}")
    print(f"loading {run_path}")
    return torch.load(run_path, map_location="cpu", weights_only=False)


def build_pde_from_bundle(bundle):
    """Reconstruct PDE from saved run bundle."""
    pde_class = bundle.get("pde_class", "HeatEquation1D")
    pde_params = bundle.get("pde_params")
    if pde_params:
        if pde_class == "HeatEquation1D":
            return HeatEquation1D(**pde_params)
        elif pde_class == "ViscousBurgers1D":
            return ViscousBurgers1D(**pde_params)
    # backward compat: old bundles without pde_class
    return HeatEquation1D(
        L=bundle.get("L", 1.0),
        T=bundle.get("T", 1.0),
        alpha=bundle.get("alpha", 1.0),
    )


def append_eval_result(name, run_id, max_abs_error, l2_error):
    os.makedirs("logs", exist_ok=True)
    write_header = not os.path.exists(EVALS_CSV)
    with open(EVALS_CSV, "a", newline="") as f:
        writer = csv.DictWriter(f, fieldnames=EVALS_FIELDNAMES)
        if write_header:
            writer.writeheader()
        writer.writerow({
            "name": name,
            "timestamp": datetime.datetime.now().strftime("%Y-%m-%dT%H:%M:%S"),
            "run_id": run_id,
            "max_abs_error": f"{max_abs_error:.6e}",
            "l2_error": f"{l2_error:.6e}",
        })
    print(f"appended to {EVALS_CSV}")


def plot_evals(names=None, title="Evaluation comparison", out_path=None):
    if not os.path.exists(EVALS_CSV):
        sys.exit(f"no evals CSV found at {EVALS_CSV} — run eval.py --name first")

    try:
        import matplotlib.pyplot as plt
        import numpy as np
    except ImportError:
        sys.exit("numpy + matplotlib required. run: pip install numpy matplotlib")

    all_rows = []
    with open(EVALS_CSV, newline="") as f:
        for row in csv.DictReader(f):
            all_rows.append({
                "name":          row["name"],
                "max_abs_error": float(row["max_abs_error"]),
                "l2_error":      float(row["l2_error"]),
            })

    if names is not None:
        name_set = set(names)
        # preserve order of names list, pick last occurrence per name
        by_name = {}
        for r in all_rows:
            if r["name"] in name_set:
                by_name[r["name"]] = r
        rows = [by_name[n] for n in names if n in by_name]
    else:
        rows = all_rows

    if not rows:
        sys.exit("evals CSV is empty")

    row_names = [r["name"] for r in rows]
    max_errs  = [r["max_abs_error"] for r in rows]
    l2_errs   = [r["l2_error"] for r in rows]
    x         = np.arange(len(rows))
    width     = 0.35

    fig, ax = plt.subplots(figsize=(max(6, len(rows) * 0.9 + 2), 5))
    ax.bar(x - width / 2, max_errs, width, label="max abs error", color="tab:red",   alpha=0.8)
    ax.bar(x + width / 2, l2_errs,  width, label="L2 error",      color="tab:blue",  alpha=0.8)
    ax.set_yscale("log")
    ax.set_xticks(x)
    ax.set_xticklabels(row_names, rotation=30, ha="right")
    ax.set_ylabel("error (log scale)")
    ax.set_title(title)
    ax.legend()
    ax.grid(True, axis="y", alpha=0.3)
    fig.tight_layout()

    os.makedirs("plots", exist_ok=True)
    out = out_path or os.path.join("plots", "evals_comparison.png")
    os.makedirs(os.path.dirname(out), exist_ok=True)
    fig.savefig(out, dpi=150)
    plt.close(fig)
    print(f"saved {out}")


def main():
    parser = argparse.ArgumentParser(description="Evaluate a saved PINN run.")
    parser.add_argument(
        "run",
        nargs="?",
        default=None,
        help="Path to run_*.pt file. Defaults to latest in runs/.",
    )
    parser.add_argument(
        "--grid",
        type=int,
        default=100,
        help="Eval grid resolution per axis (default 100).",
    )
    parser.add_argument(
        "--name",
        default=None,
        help="Label for this evaluation (required to log results to evals.csv).",
    )
    parser.add_argument(
        "--plot",
        action="store_true",
        help="Plot evals.csv comparison bar chart instead of running evaluation.",
    )
    args = parser.parse_args()

    if args.plot:
        plot_evals()
        return

    if args.name is None:
        sys.exit("--name is required. e.g.: python eval.py --name baseline")

    run_path = resolve_run_path(args.run)
    bundle = load_run(run_path)
    run_id = bundle.get("run_id", os.path.splitext(os.path.basename(run_path))[0].replace("run_", ""))

    model = PINN(**bundle["model_config"])
    model.load_state_dict(bundle["model_state"])

    pde = build_pde_from_bundle(bundle)
    result = evaluate(model, pde, device="cpu", grid=args.grid)

    if "max_abs_error" in result:
        print(f"max abs error: {result['max_abs_error']:.4e}")
        print(f"L2 error:      {result['l2_error']:.4e}")
        append_eval_result(args.name, run_id, result["max_abs_error"], result["l2_error"])
    else:
        print("no analytical solution — prediction only, nothing logged")


if __name__ == "__main__":
    main()

File: cpp/src/eval.cpp
#include "eval.hpp"
#include <algorithm>
#include <chrono>
#include <cstdio>
#include <filesystem>
#include <fstream>
#include <iomanip>
#include <iostream>
#include <sstream>
#include <stdexcept>
#include <vector>

namespace fs = std::filesystem;

static const std::string EVALS_CSV = "c_logs/evals.csv";
static const std::string EVALS_HEADER = "name,timestamp,run_id,max_abs_error,l2_error\n";

EvalResult evaluate(PINN& model, PDEProblem& pde, torch::Device dev, int grid) {
    model.eval();
    torch::NoGradGuard no_grad;

    auto xs = torch::linspace(pde.x_min, pde.x_max, grid,
                               torch::TensorOptions().dtype(torch::kFloat32).device(dev));
    auto ts = torch::linspace(0.0, pde.t_max, grid,
                               torch::TensorOptions().dtype(torch::kFloat32).device(dev));

    auto grids = torch::meshgrid({xs, ts}, "ij");
    auto X  = grids[0];
    auto Tm = grids[1];

    auto x_flat = X.reshape({-1, 1});
    auto t_flat = Tm.reshape({-1, 1});

    auto u_pred = model.forward(x_flat, t_flat);

    EvalResult result;
    result.u_pred = u_pred;
    result.X = X;
    result.Tm = Tm;

    auto u_exact_opt = pde.analytical_solution(x_flat, t_flat);
    if (u_exact_opt.has_value()) {
        auto& u_exact = u_exact_opt.value();
        auto diff = u_pred - u_exact;
        result.u_exact = u_exact;
        result.max_abs_error = diff.abs().max().item<double>();
        result.l2_error = (diff.norm() / u_exact.norm()).item<double>();
    }

    model.train();
    return result;
}

void append_eval_result(const std::string& name, const std::string& run_id,
                         double max_abs_error, double l2_error) {
    fs::create_directories("c_logs");

    bool write_header = !fs::exists(EVALS_CSV);
    std::ofstream f(EVALS_CSV, std::ios::app);
    if (!f) throw std::runtime_error("cannot open " + EVALS_CSV);

    if (write_header) f << EVALS_HEADER;

    // timestamp YYYY-MM-DDTHH:MM:SS
    auto now = std::chrono::system_clock::now();
    auto tt  = std::chrono::system_clock::to_time_t(now);
    std::tm tm_buf{};
#ifdef _WIN32
    localtime_s(&tm_buf, &tt);
#else
    localtime_r(&tt, &tm_buf);
#endif
    std::ostringstream ts;
    ts << std::put_time(&tm_buf, "%Y-%m-%dT%H:%M:%S");

    char max_buf[32], l2_buf[32];
    std::snprintf(max_buf, sizeof(max_buf), "%.6e", max_abs_error);
    std::snprintf(l2_buf,  sizeof(l2_buf),  "%.6e", l2_error);

    f << name << "," << ts.str() << "," << run_id << ","
      << max_buf << "," << l2_buf << "\n";

    std::cout << "appended to " << EVALS_CSV << "\n";
}

std::string resolve_run_path(const std::optional<std::string>& arg) {
    if (arg.has_value()) return arg.value();

    std::vector<std::string> candidates;
    if (fs::exists("c_runs")) {
        for (auto& p : fs::directory_iterator("c_runs")) {
            auto s = p.path().string();
            if (s.find("run_") != std::string::npos &&
                p.path().extension() == ".pt") {
                candidates.push_back(s);
            }
        }
    }
    if (candidates.empty())
        throw std::runtime_error("no runs found in c_runs/ — pass a path or run pinn_experiments first");

    std::sort(candidates.begin(), candidates.end());
    return candidates.back();
}

std::unique_ptr<PDEProblem> build_pde_from_class(const std::string& pde_class) {
    if (pde_class == "ViscousBurgers1D")
        return std::make_unique<ViscousBurgers1D>();
    return std::make_unique<HeatEquation1D>();
}
